WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m

```diff
@@ -24,37 +32,123 @@ - (void)dealloc
-                        hint:(NSString *)hint
-                  isPassword:(BOOL)isPassword
+- (instancetype)initWithStyle:(UITableViewStyle)style
+{
+    return [self initWithText:@"" placeholder:@"" hint:@""];
+}
+
+- (instancetype)initWithText:(NSString *)text placeholder:(NSString *)placeholder hint:(NSString *)hint
 {
     self = [super initWithStyle:UITableViewStyleGrouped];
     if (self) {
         _text = text;
         _placeholder = placeholder;
         _hint = hint;
-        _isPassword = isPassword;
     }
     return self;
 }
 
-- (instancetype)initWithStyle:(UITableViewStyle)style
+
+#pragma mark - View Lifecycle
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+    [WPStyleGuide resetReadableMarginsForTableView:self.tableView];
+    [WPStyleGuide configureColorsForView:self.view andTableView:self.tableView];
+}
+
+- (void)viewWillAppear:(BOOL)animated
 {
-    return [self initWithText:@"" placeholder:@"" hint:@"" isPassword:NO];
+    [super viewWillAppear:animated];
+    [self setupModalButtonsIfNeeded];
 }
 
 - (void)viewDidAppear:(BOOL)animated
 {
-    [self.textField becomeFirstResponder];
     [super viewDidAppear:animated];
+    [self.textField becomeFirstResponder];
 }
 
-- (void)viewDidLoad
+- (void)viewWillDisappear:(BOOL)animated
 {
-    [super viewDidLoad];
-    [WPStyleGuide resetReadableMarginsForTableView:self.tableView];
-    [WPStyleGuide configureColorsForView:self.view andTableView:self.tableView];
+    [super viewWillDisappear:animated];
+
+    [self.view endEditing:YES];
+    
+    if (self.onValueChanged && ![self.textField.text isEqualToString:self.text] && self.textPassesValidation) {
+        self.onValueChanged(self.textField.text);
+    }
+}
+
+
+#pragma mark - Helpers
+
+- (void)setupModalButtonsIfNeeded
+{
+    // Proceed only if this is the only VC in the current navigationController
+    if (self.isBeingPresented == true || self.isModal == false) {
+        return;
+    }
+    
+    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
+                                                                                          target:self
+                                                                                          action:@selector(cancelButtonWasPressed:)];
+    
+    
+    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
+                                                                                           target:self
+                                                                                           action:@selector(doneButtonWasPressed:)];
+}
+
+- (IBAction)cancelButtonWasPressed:(id)sender
+{
+    [self dismissViewControllerAnimated:true completion:nil];
+}
+
+- (IBAction)doneButtonWasPressed:(id)sender
+{
+    if (self.textPassesValidation == false) {
+        [self displayValidationAlert];
```

Comment:
Instead of an alert, would it be better to disable the Done button when the text isn't valid?